dbms/src/Storages/DeltaMerge/ColumnFile/ColumnFileTinyVectorIndexWriter.cpp

```diff
@@ -0,0 +1,234 @@
+// Copyright 2024 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <IO/Compression/CompressedWriteBuffer.h>
+#include <Storages/DeltaMerge/ColumnFile/ColumnFileTiny.h>
+#include <Storages/DeltaMerge/ColumnFile/ColumnFileTinyReader.h>
+#include <Storages/DeltaMerge/ColumnFile/ColumnFileTinyVectorIndexWriter.h>
+#include <Storages/DeltaMerge/Index/VectorIndex.h>
+#include <Storages/DeltaMerge/dtpb/dmfile.pb.h>
+
+
+namespace DB::ErrorCodes
+{
+extern const int ABORTED;
+} // namespace DB::ErrorCodes
+
+namespace DB::DM
+{
+
+ColumnFileTinyVectorIndexWriter::LocalIndexBuildInfo ColumnFileTinyVectorIndexWriter::getLocalIndexBuildInfo(
+    const LocalIndexInfosSnapshot & index_infos,
+    const ColumnFilePersistedSetPtr & file_set)
+{
+    assert(index_infos != nullptr);
+
+    LocalIndexBuildInfo build;
+    build.indexes_to_build = std::make_shared<LocalIndexInfos>();
+    build.file_ids.reserve(file_set->getColumnFileCount());
+    for (const auto & file : file_set->getFiles())
+    {
+        auto * tiny_file = file->tryToTinyFile();
+        if (!tiny_file)
+            continue;
+
+        bool any_new_index_build = false;
+        for (const auto & index : *index_infos)
+        {
+            auto schema = tiny_file->getSchema();
+            assert(schema != nullptr);
+            // The ColumnFileTiny may be built before col_id is added. Skip build indexes for it
+            if (!schema->getColIdToOffset().contains(index.column_id))
+                continue;
+
+            // Index already built, skip
+            if (tiny_file->hasIndex(index.index_id))
+                continue;
+
+            any_new_index_build = true;
+            // FIXME: the memory usage is not accurate, but it's fine for now.
+            build.estimated_memory_bytes += tiny_file->getBytes();
+
+            // Avoid duplicate index build
+            if (std::find(build.index_ids.begin(), build.index_ids.end(), index.index_id) == build.index_ids.end())
+            {
+                build.indexes_to_build->emplace_back(index);
+                build.index_ids.emplace_back(index.index_id);
```

Comment:
What about set `any_new_index_build` and accumulate `estimated_memory_bytes` only in this if branch?

Reply:
Check the reply before: https://github.com/pingcap/tiflash/pull/9546#discussion_r1815954981